### PR TITLE
Make base64 functions isomorphic

### PR DESCRIPTION
**Pull request by nwalters512**
# Description

~Back when these functions were first introduced, Node didn't support `btoa` or `atob`. Since v16, it does! This allows us to rewrite the functions to work the same everywhere.~

Since originally authoring this, we ended up refactoring to use the `Buffer` global when it's available (in Node) and fall back to `btoa` and `atob` when it's not (in the browser). This allows us to take advantage of higher performance on Node.

# Testing

I'll let the automated tests do their thing.

### CODE REVIEW

File: apps/prairielearn/src/lib/base64-util.ts [hunk 1]
```diff
@@ -1,37 +1,22 @@
...
-  try {
-    return btoa(
-      encodeURIComponent(str).replaceAll(/%([0-9A-F]{2})/g, (match, p1) => {
-        return String.fromCharCode(Number.parseInt('0x' + p1, 16));
-      }),
-    );
-  } catch {
-    logger.error(`b64EncodeUnicode: returning empty string because failed to encode ${str}`);
-    return '';
-  }
```
Comment (nwalters512) [inline apps/prairielearn/src/lib/base64-util.ts:19]: Logging and returning an empty string is _very_ strange behavior. This actually bit me while working on agentic question generation. I accidentally passed in a non-base64-encoded string to the decode function and was very surprised when I was getting back an empty string. It wasn't until I looked at the logs that I realized what had happened.

IMO it's better to just throw an error if this fails than somewhat-silently using an empty string.

File: apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts [hunk 2]
```diff
@@ -26,6 +26,7 @@ import {
   mergeUsage,
 } from '../../lib/ai.js';
 import * as b64Util from '../../lib/base64-util.js';
+import { b64EncodeUnicode } from '../../lib/base64-util.js';
```
Comment (jonatanschroeder) [inline apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts]: You probably don't need both?

Comment (nwalters512) [reply apps/prairielearn/src/ee/lib/aiQuestionGeneration.ts]: Yep, CI caught this, resolved in b16692e.

File: apps/prairielearn/src/ee/pages/instructorAiGenerateDraftEditor/instructorAiGenerateDraftEditor.ts [hunk 3]
```diff
@@ -7,6 +7,7 @@ import { flash } from '@prairielearn/flash';
 import { execute, loadSqlEquiv, queryRow, queryRows } from '@prairielearn/postgres';
 
 import * as b64Util from '../../../lib/base64-util.js';
```
Comment (jonatanschroeder) [inline apps/prairielearn/src/ee/pages/instructorAiGenerateDraftEditor/instructorAiGenerateDraftEditor.ts]: One more instance of the same issue (though CI likely didn't catch this because it's still in use). Same in other files below.

**Comment by reteps** [comment]
The docs for `atob` and `btoa` say: https://nodejs.org/api/globals.html#atobdata

> Legacy. Use Buffer.from(data, 'base64') instead. This function is only provided for compatibility with legacy web platform APIs and should never be used in new code, because they use strings to represent binary data and predate the introduction of typed arrays in JavaScript

Do we care about this?

**Comment by jonatanschroeder** [comment]
> The docs for `atob` and `btoa` say: https://nodejs.org/api/globals.html#atobdata
> 
> > Legacy. Use Buffer.from(data, 'base64') instead. This function is only provided for compatibility with legacy web platform APIs and should never be used in new code, because they use strings to represent binary data and predate the introduction of typed arrays in JavaScript
> 
> Do we care about this?
> 

Buffer is specific to node, not available client-side.

File: apps/prairielearn/src/ee/pages/instructorAiGenerateDraftEditor/instructorAiGenerateDraftEditor.ts (hunk 3 above)
Comment (nwalters512) [reply apps/prairielearn/src/ee/pages/instructorAiGenerateDraftEditor/instructorAiGenerateDraftEditor.ts]: Fixed in 84302b1. I cleaned up a few more while I was at it.

File: apps/prairielearn/src/lib/base64-util.ts (hunk 1 above)
Comment (mwest1066) [reply apps/prairielearn/src/lib/base64-util.ts:19]: Agreed

File: apps/prairielearn/src/lib/base64-util.test.ts [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    { name: 'Russian characters', value: 'Привет мир' },
+
+    // Emoji and mixed
+    { name: 'emoji', value: '🎉🎊🎈' },
+    { name: 'mixed scripts', value: 'Mixed: hello 世界 مرحبا' },
+
+    // Edge cases
+    { name: 'empty string', value: '' },
+    { name: 'single character', value: 'a' },
+  ];
```
Comment (nwalters512) [inline apps/prairielearn/src/lib/base64-util.test.ts:44]: Claude Sonnet 4.5 wrote these cases, I massaged them into the tests below.

File: apps/prairielearn/src/lib/base64-util.test.ts [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    { name: 'simple ASCII string', value: 'hello' },
+    { name: 'string with punctuation', value: 'Hello, World!' },
+    { name: 'numeric string', value: '123456789' },
+
+    // Whitespace and escapes
+    { name: 'whitespace characters', value: '\n\t\r' },
+    { name: 'escaped whitespace strings', value: '\\n\\t\\r' },
+
+    // Symbols and specials
+    { name: 'special characters', value: 'special!@#$%^&*()' },
```
Comment (jonatanschroeder) [inline apps/prairielearn/src/lib/base64-util.test.ts]: Nit: I'd suggest adding characters that have special meaning in HTML, JS or URLs, for extra precaution.

```suggestion
    { name: 'special characters', value: 'special!@#$%^&*()<>?"\'`=+\\' },
```
